### PR TITLE
refactor: Improve the error messages from prepare.rs

### PR DESCRIPTION
**Pull request by nagisa**
Contract preparation is the first code that will typically see the
contract code and consequentially it is also the first which will report
issues with the contract.

One such example is #5310 for which the preparation code should now
report something more meaningful.

I went ahead and implemented `std::error::Error` in some places, but it
seems to me like we probably want to move the entire file to
`thiserror`...

This is indended to be NFC in that there should not be any meaningful functionality
differences other than a change the error messages and variants seen.

### CODE REVIEW

**Comment by nagisa** [comment]
Some observations for future work: we should not be formatting the causations into the `Display`ed error message. Instead all of the errors should have a reasonable `source` chain and the error messages should be formatted into the `: ` separated form at the presentation layer or similar, rather than baked into the code.

File: runtime/near-vm-errors/src/lib.rs [hunk 1]
```diff
@@ -130,9 +154,10 @@ pub enum CompilationError {
 /// Error that can occur while preparing or executing Wasm smart-contract.
 pub enum PrepareError {
     /// Error happened while serializing the module.
-    Serialization,
+    Serialization(StringError),
     /// Error happened while deserializing the module.
-    Deserialization,
+    Deserialization(StringError),
```
Comment (matklad) [inline runtime/near-vm-errors/src/lib.rs:159]: >This is indended to be NFC in that there should not be any meaningful functionality
differences other than a change the error messages and variants seen.

Alas, that's not the case, and this is one example. The following code works on master, but fails with this PR:

```rust
#[test]
fn serialization_compat() {
    let _: PrepareError = BorshDeserialize::try_from_slice(&[1]).unwrap();
}
```

And this `[1]` is stored in the database for real transactions.

**Comment by matklad** [comment]
>separated form at the presentation layer or similar, rather than baked into the code.

Sadly, this *is* the presentation layer at the moment. Specifically, `#[derive(Serialize)]` is what defines our HTTP API (relevant entry struct: https://github.com/near/nearcore/blob/feb0acdb8b798d77e10b2217bcaa9720f93557f8/core/primitives/src/views.rs#L1040-L1063). 

You can see how the user sees the errors here: https://explorer.testnet.near.org/transactions/45RShTh6wCnRg3PQVu9nBPXZ7PsK5hqFyEGPEAfCNX7n#GGs6dB4ktqJhVNN4gZBPNoKZiPAw4VSSqExouHsFGuNw

**Comment by nagisa** [comment]
Yeah, okay, never mind then. I can see approaches we could take to make changes backwards compatibly, but I don't really see it being worth it without a more in-depth solution to decouple code errors and database/API layers.

**Pull request closed without merging**